awslabs/aws-lambda-powertools-python
# feat(parser): add support for Pydantic v2

**Issue number:** #2672 
Additional issue: #2427 

## Summary

### Changes

After an extensive week of investigation, experimentation, and implementation, I'm excited to open this pull request, which introduces support for Pydantic v2 in the parser feature. I wanted to provide you with an overview of the changes and considerations made during this process:

**Support for Pydantic v1 and v2:**

- The parser utility has been enhanced to ensure compatibility with both Pydantic v1 and v2.
- Users now have the flexibility to choose their preferred version of Pydantic for parse utility.

**Powertools and the support for Pydantic:**

- Historically, Powertools has considered Pydantic v1 as the official library for parse utility, given its long-standing compatibility and robustness within our code.
- However, with this update, we are embracing the advancements and enhancements introduced in Pydantic v2, allowing users to leverage the latest features and improvements.
- It's important to highlight that while we have introduced support for Pydantic v2, our official version for Powertools remains Pydantic v1. This decision has been made to prioritize compatibility and provide a seamless experience for our existing users. By maintaining Pydantic v1 as our official library, we ensure that our ecosystem and the applications built with it continue to function as expected without any disruptions.


**No-breaking changes were made**

**NOTE**: We haven't detected any breaking changes, but please let us know as soon as possible if you find any.

- Importantly, this update does not introduce any breaking changes to the existing functions within the Powertools framework. The functions that were already in use remain intact and fully functional, ensuring compatibility with the existing codebase. 
- However, it's worth noting that with the support for Pydantic v2, some Pydantic functions used in Powertools are now considered deprecated. This deprecation serves as a signal to users that these functions will be removed in a future major release, specifically Powertools v3.

#### Compatibility table

| Pydantic v1     | Pydantic v2           | V2 deprecation     | In use?            | Code change required |
| --------------- | --------------------- | ------------------ | ------------------ | -------------------- |
| @validator      | @field_validator      | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:   |
| @root_validator | @model_validator      | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:   |
| .parse_obj()    | model_validate()      | :heavy_check_mark: | :heavy_check_mark: |                      |
| .json()         | model_dump_json()     | :heavy_check_mark: | :heavy_check_mark: |                      |
| .dict()         | model_dump()          | :heavy_check_mark: | :heavy_check_mark: |                      |
| .parse_raw()    | model_validate_json() | :heavy_check_mark: | :heavy_check_mark: |                      |


- To ensure a smooth transition and minimize disruptions for our users, we have temporarily suppressed the `PydanticDeprecatedSince20` and `PydanticDeprecationWarning` warnings (related to these functions). This allows existing applications to continue functioning as expected without outputting warnings.
- If needed, you can enable the warnings yourself with something like the code below. Reference: https://docs.python.org/3/library/warnings.html
```python
from aws_lambda_powertools.utilities.parser import event_parser, BaseModel, envelopes
from aws_lambda_powertools.utilities.parser.models import (
    SqsModel,
)

from aws_lambda_powertools import Logger
import pydantic

import warnings
warnings.simplefilter('default')

```

**How to test before we merge this PR**

To facilitate the testing of Pydantic v2 before merging this pull request, we have created a separate branch called `poc/pydanticv2` where you can find the relevant code changes. Here are the steps to follow in order to test it effectively:

If using sourcecode:
	- Clone Powertools repository
	- Fetch the `poc/pydanticv2` branch
	- Check out the branch locally by running the command: `git checkout poc/pydanticv2`.

If using pip:

Create or modify the `requirements.txt` file and add the following content:
```
git+https://github.com/aws-powertools/powertools-lambda-python@poc/pydanticv2#egg=aws_lambda_powertools
pydantic>2
```
Then install by running: `pip install -r requirements.txt`.

If using Poetry:

```sh
poetry add git+https://github.com/aws-powertools/powertools-lambda-python#poc/pydanticv2
poetry add pydantic>2
```

### Note for documentation

Since the Powertools currently relies on Pydantic v1 as its official dependency, you need to change the `aws-lambda-powertools[tracer]` or `aws-lambda-powertools[all]` installation to `aws-lambda-powertools` or `aws-lambda-powertools[validation,tracer,aws-sdk]` and add Pydantic v2 as an extra dependency.

### Missing in this PR
- [x] Workflow in develop branch to ensure that any changes will work in Pydanticv2
- [x] Community feedback

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.
